### PR TITLE
Egork/figure table predictors

### PR DESCRIPTION
**Pull request by comorado**

### CODE REVIEW

**Comment by soldni** [comment]
Quickly changed CI tests config so that `test_figure_table_predictors.py` runs alongside the other tests requiring VILA. PR looks good to me otherwise!

**Comment by soldni** [comment]
Hey @comorado , some tests seem to be failing after recent commits?

**Review by cmwilhelm — CHANGES_REQUESTED** [review]
Most of my comments are related to taking a dependency on vila and layout parser directly in your predictor.

All the other predictors assume their upstream dependencies have already been run, and that they are just operating over their outputs. So your predictor shouldn't need to instantiate Vila and LayoutParser itself.

Likewise, we shouldn't encode that dependency in the timo config -- it adds a huge bloat for a model that is just doing some analysis on top of provided span groups.

**Comment by comorado** [comment]
@soldni Thank you, yes it seems that it is because of this [change](https://github.com/allenai/mmda/pull/150/files) see my comment.

File: ai2_internal/config.yaml [hunk 1]
```diff
@@ -250,3 +250,36 @@ model_variants:
...
+    predictor: ai2_internal.figure_table_predictors.interface.Predictor
+
+    # Class path to pydantic PredictorConfig implementation in <model_package_name>==<model_package_version>
+    predictor_config: ai2_internal.bibentry_detection_predictor.interface.PredictorConfig
+
+    integration_test: ai2_internal.figure_table_predictors.integration_test.TestInterfaceIntegration
+
+    # Any additional sets of dependencies required by the model.
+    # These are the 'extras_require' keys in your setup.py.
+    extras_require: ["dev", "vila_predictors"]
```
[the remark this answers is not in the document]
Comment (comorado) [reply ai2_internal/config.yaml:271]: Yes that is correct, thank you for pointing this out.

File: ai2_internal/config.yaml [hunk 2]
```diff
@@ -250,3 +250,36 @@ model_variants:
...
+
+    # Full S3 path to tar.gz'ed artifacts archive, nullable
+    # https://huggingface.co/allenai/ivila-row-layoutlm-finetuned-s2vl-v2
+    artifacts_s3_path: s3://ai2-timo-registry/model-artifacts/vila/ivila-row-layoutlm-finetuned-s2vl-v2.tar.gz
```
[the remark this answers is not in the document]
Comment (comorado) [reply ai2_internal/config.yaml:275]: That is right

File: mmda/predictors/heuristic_predictors/figure_table_predictors.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+from mmda.types.span import Span
+from mmda.predictors.hf_predictors.token_classification_predictor import IVILATokenClassificationPredictor
+from mmda.parsers.pdfplumber_parser import PDFPlumberParser
+from mmda.rasterizers.rasterizer import PDF2ImageRasterizer
+from mmda.predictors.lp_predictors import LayoutParserPredictor
+from mmda.utils.tools import MergeSpans
+
+
+class FigureTablePredictions(BaseHeuristicPredictor):
+    REQUIRED_BACKENDS = ['layoutparser', 'vila']
```
[the remark this answers is not in the document]
Comment (comorado) [reply mmda/predictors/heuristic_predictors/figure_table_predictors.py:22]: Correct, will take this dependencies down.

File: mmda/predictors/heuristic_predictors/figure_table_predictors.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    def predict(self, document: Document) -> List[SpanGroup]:
+        """
+        Predict figure->caption mapping and, table->caption mapping.
+        Return: List[SpanGroup], SpanGroup has start, end corresponding to caption start, end indexes and box
+        corresponding to merged boxes of the tokens of the caption. Type is one of ['Figure', 'Table']. BoxGroup stores
+        information about the boundaries of figure or table.
+        """
+        predictions = []
+        predictions.extend(self._predict(document, caption_type='Figure'))
+        predictions.extend(self._predict(document, caption_type='Table'))
```
[the remark this answers is not in the document]
Comment (comorado) [reply mmda/predictors/heuristic_predictors/figure_table_predictors.py:174]: @cmwilhelm I need a single list, each of _predict returns a list so probably
`list(zip(self._predict(document, caption_type="Figure"), self._predict(document, caption_type="Table")))`

File: mmda/predictors/heuristic_predictors/figure_table_predictors.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+from mmda.predictors.hf_predictors.token_classification_predictor import IVILATokenClassificationPredictor
+from mmda.parsers.pdfplumber_parser import PDFPlumberParser
+from mmda.rasterizers.rasterizer import PDF2ImageRasterizer
+from mmda.predictors.lp_predictors import LayoutParserPredictor
+from mmda.utils.tools import MergeSpans
+
+
+class FigureTablePredictions(BaseHeuristicPredictor):
+    REQUIRED_BACKENDS = ['layoutparser', 'vila']
+    REQUIRED_DOCUMENT_FIELDS = ['layoutparser_span_groups', 'vila_span_groups']
```
[the remark this answers is not in the document]
Comment (comorado) [reply mmda/predictors/heuristic_predictors/figure_table_predictors.py:23]: Images is not part of the fields for some reason. We do not add it to the [self._fields](https://github.com/allenai/mmda/blob/main/mmda/types/document.py#L102-L122)

File: ai2_internal/figure_table_predictors/interface.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        doc.annotate(layoutparser_span_groups=[sg.to_mmda() for sg in self.layoutparser_span_groups])
+        doc.annotate(vila_span_groups=[sg.to_mmda() for sg in self.vila_span_groups])
+        return doc
+
+
+class Prediction(BaseModel):
+    """
+    Describes the outcome of inference for one Instance
+    """
+    table_figure_caption_list: List[api.SpanGroup]
```
[the remark this answers is not in the document]
Comment (comorado) [reply ai2_internal/figure_table_predictors/interface.py:54]: Hmm I see so you proposing to return a tuple we can definitely do that!

File: setup.py [hunk 7]
```diff
@@ -15,7 +15,8 @@
         "pandas",
         "pydantic",
         "ncls",
-        "necessary"
+        "necessary",
+        "scipy",
```
[the remark this answers is not in the document]
Comment (comorado) [reply setup.py:19]: @cmwilhelm I've tried to move scipy into `extra_requires` section but getting an error
`/usr/local/lib/python3.8/site-packages/mmda/predictors/heuristic_predictors/figure_table_predictors.py:5: in <module>
    from scipy.optimize import linear_sum_assignment
E   ModuleNotFoundError: No module named 'scipy'
`

Here is the lists I added
`"figure_table_predictors": ["scipy"],`

File: ai2_internal/figure_table_predictors/interface.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        doc.annotate(vila_span_groups=[sg.to_mmda() for sg in self.vila_span_groups])
+        return doc
+
+
+class Prediction(BaseModel):
+    """
+    Describes the outcome of inference for one Instance
+    """
+    table_figure_list: List[api.SpanGroup]
+    table_caption_list: List[api.SpanGroup]
```
Comment (cmwilhelm) [inline ai2_internal/figure_table_predictors/interface.py:55]:
```python
table_list: List[api.SpanGroup]
figure_list: List[api.SpanGroup]
```

?

**Review by cmwilhelm — APPROVED** [review]
LGTM except one small question on the output `Instance`. Looks like the field names might be off?

Thanks for the extra round of attention here.

**Pull request merged**